core/pta/remote_attestation/hash.c

```diff
@@ -0,0 +1,123 @@
+ * (ldelf, kernel or temporary mappings).
+ */
+static bool is_region_valid(struct vm_region *r)
+{
+    uint32_t dontwant = VM_FLAG_EPHEMERAL | VM_FLAG_PERMANENT |
+                        VM_FLAG_LDELF;
+    uint32_t want = VM_FLAG_READONLY;
+
+    return ((r->flags & want) == want && !(r->flags & dontwant));
+}
+
+/*
+ * With this comparison function, we're hashing the smaller regions first.
+ * Regions of equal size are ordered based on their content (memcmp()).
+ * Identical regions can be in any order since they will yield the same hash
+ * anyways.
+ */
+static int cmp_regions(const void *a, const void *b) {
+    const struct vm_region *r1 = *(const struct vm_region **)a;
+    const struct vm_region *r2 = *(const struct vm_region **)b;
+
+    if (r1->size < r2->size)
+        return -1;
+
+    if (r1->size > r2->size)
+        return 1;
+
+    return memcmp((void *)r1->va, (void *)r2->va, r1->size);
+}
+
+static TEE_Result hash_regions(struct vm_info *vm_info, uint8_t *hash) {
+    TEE_Result res = TEE_SUCCESS;
+    struct vm_region *r = NULL;
+    struct vm_region **regions = NULL;
+    size_t nregions = 0;
+    void *ctx = NULL;
+    size_t i = 0;
+
+    res = crypto_hash_alloc_ctx(&ctx, TEE_ALG_SHA256);
+    if (res)
+        return res;
+
+    res = crypto_hash_init(ctx);
+    if (res)
+        goto out;
+
+    /*
+     * Make an array of region pointers so we can use qsort() to order it.
+     */
+
+    TAILQ_FOREACH(r, &vm_info->regions, link)
+    if (is_region_valid(r))
+        nregions++;
+
+    regions = malloc(nregions * sizeof(*regions));
+    if (!regions) {
+        res = TEE_ERROR_OUT_OF_MEMORY;
+        goto out;
+    }
+
+    TAILQ_FOREACH(r, &vm_info->regions, link)
+    if (is_region_valid(r))
+        regions[i++] = r;
+
+    /*
+     * Sort regions so that they are in a consistent order even when TA ASLR
+     * is enabled.
+     */
+    qsort(regions, nregions, sizeof(*regions), cmp_regions);
+
+    /* Hash regions in order */
+    for (i = 0; i < nregions; i++) {
+        r = regions[i];
+        DMSG("va %p size %zu", (void *)r->va, r->size);
+        res = crypto_hash_update(ctx, (uint8_t *)r->va, r->size);
+        if (res)
+            goto out;
+    }
+
+    res = crypto_hash_final(ctx, hash, TEE_SHA256_HASH_SIZE);
+out:
+    if (regions)
+        free(regions);
+    if (ctx)
+        crypto_hash_free_ctx(ctx);
+    return res;
+}
+
+TEE_Result get_hash_ta_memory(uint8_t *out, size_t out_sz) {
+    struct user_mode_ctx *uctx = NULL;
+    TEE_Result res = TEE_SUCCESS;
+    struct ts_session *s = NULL;
+
+    /* Check that we're called from a user TA */
+    s = ts_get_calling_session();
+    if (!s)
+        return TEE_ERROR_ACCESS_DENIED;
+    uctx = to_user_mode_ctx(s->ctx);
+    if (!uctx)
+        return TEE_ERROR_ACCESS_DENIED;
+
+    /* The output buffer must be large enough to hold the hash. */
+    if (out_sz < TEE_SHA256_HASH_SIZE)
```

Comment:
Nitpicking: this could be removed by expecting a well sized buffer:
`TEE_Result get_hash_ta_memory(uint8_t out[TEE_SHA256_HASH_SIZE]);

Reply:
DONE